Detections/AzureDevOps/AzDOHistoricServiceConnectionAdds.yaml

```diff
@@ -0,0 +1,44 @@
+id: 5efb0cfd-063d-417a-803b-562eae5b0301
+name: Azure DevOps Service Conection Addition/Abuse - Historic Allowlist
+description: |
+  'This detection builds a allowlist of historic service connection use by Builds and Releases and compares to recent history, flagging growth of service connection use in non manually allowlisted, non historically allowlisted Build/Release runs.
+  This is to determine if someone is hijacking a build/release and adding many service connections in order to abuse or dump credentials from service connections.'
+severity: High
+queryFrequency: 2h
+queryPeriod: 14d
+triggerOperator: gt
+triggerThreshold: 0
+tactics:
+  - Execution
+  - Impact
+  - Escalation
+query: |
+
+  // Ignore Build/Releases with less/equal this number
+  let ServiceConnectionThreshold = 3;
+  // New Connections need to exhibit execution of more "new" connections than this number.
+  let NewConnectionThreshold = 1;
+  // List of Builds/Releases to ignore
+  let BypassDefIds = datatable(DefId:string, Type:string, ProjectName:string)
+  [
+  //"103", "Release", "ProjectA",
+  //"42", "Release", "ProjectB",
+  //"122", "Build", "ProjectB"
+  ];
+  let HistoricDefs = AzureDevOpsAuditing
+  | where TimeGenerated > ago(14d) and TimeGenerated < ago(6h)
+  | where OperationName == "Library.ServiceConnectionExecuted" 
+  | extend DefId = tostring(Data.DefinitionId), Type = tostring(Data.PlanType), ConnectionId = tostring(Data.ConnectionId)
+  | summarize HistoricCount = dcount(tostring(ConnectionId)), ConnectionNames = make_set(tostring(Data.ConnectionName)) by tostring(DefId), tostring(Type), ProjectId, ProjectName;
+  AzureDevOpsAuditing
+  | where TimeGenerated >= ago(6h)
+  | where OperationName == "Library.ServiceConnectionExecuted" 
+  | extend DefId = tostring(Data.DefinitionId), Type = tostring(Data.PlanType), ConnectionId = tostring(Data.ConnectionId)
+  | parse ScopeDisplayName with OrganizationName ' (Organization)'
+  | summarize CurrentCount = dcount(tostring(ConnectionId)), ConnectionNames = make_set(tostring(Data.ConnectionName)) by OrganizationName, tostring(DefId), tostring(Type), ProjectId, ProjectName
+  | where CurrentCount > ServiceConnectionThreshold
+  | join (HistoricDefs) on ProjectId, DefId, Type
+  | join kind=anti BypassDefIds on $left.DefId==$right.DefId and $left.Type == $right.Type and $left.ProjectName == $right.ProjectName
+  | extend link = iif(Type == "Build", strcat('https://dev.azure.com/', OrganizationName, '/', ProjectName, '/_build?definitionId=', DefId) ,strcat('https://dev.azure.com/', OrganizationName, '/', ProjectName, '/_release?_a=releases&view=mine&definitionId=', DefId))
+  | where CurrentCount >= HistoricCount + NewConnectionThreshold
+  | project OrganizationName, ProjectName, DefId, link, RecentDistinctServiceConnections = CurrentCount, HistoricDistinctServiceConnections = HistoricCount, RecentConnections = ConnectionNames, HistoricConnections = ConnectionNames1
```

Comment:
Detections require that we output all custom entity types currently available in the UX - timestamp, AccountCustomEntity, HostCustomEntity, IPCustomEntity. In this case, maybe only timestamp, but wondering if we should still populate AccountCustomEntity with ActorUPN even though it seems like it isn't really a UPN.

| extend timestamp = TimeGenerated, AccountCustomEntity = ActorUPN

Reply:
theres not really a valid time generated, custom entity, etc.  the offending item is an Azure DevOps CI/CD pipeline that is abusing service connections.  not sure what to put here.